url7b90c1010b_DastanIqbal_HelloZMQwithJava_tgz-pJ8-ClientJ8/src/zmq/Mtrie.java

```diff
@@ -0,0 +1,394 @@
+        int new_max = min;
+        for (int c = 0; c != count; c++) {
+            buff_[buffsize_] = (byte) (min + c);
+            if (next[c] != null) {
+                next[c].rm_helper (pipe_, buff_, buffsize_ + 1,
+                    maxbuffsize_, func_, arg_);
+
+                //  Prune redundant nodes from the mtrie
+                if (next[c].is_redundant ()) {
+                    next[c] = null;
+
+                    assert (live_nodes > 0);
+                    --live_nodes;
+                }
+                else {
+                    //  The node is not redundant, so it's a candidate for being
+                    //  the new min/max node.
+                    //
+                    //  We loop through the node array from left to right, so the
+                    //  first non-null, non-redundant node encountered is the new
+                    //  minimum index. Conversely, the last non-redundant, non-null
+                    //  node encountered is the new maximum index.
+                    if (c + min < new_min)
+                        new_min = c + min;
+                    if (c + min > new_max)
+                        new_max = c + min;
+                }
+            }
+        }
+
+        assert (count > 1);
+
+        //  Free the node table if it's no longer used.
+        if (live_nodes == 0) {
+            next = null;
+            count = 0;
+        }
+        //  Compact the node table if possible
+        else if (live_nodes == 1) {
+            //  If there's only one live node in the table we can
+            //  switch to using the more compact single-node
+            //  representation
+            assert (new_min == new_max);
+            assert (new_min >= min && new_min < min + count);
+            Mtrie node = next [new_min - min];
+            assert (node != null);
+            next = null;
+            next = new Mtrie[]{node};
+            count = 1;
+            min = new_min;
+        }
+        else if (live_nodes > 1 && (new_min > min || new_max < min + count - 1)) {
+            assert (new_max - new_min + 1 > 1);
+
+            Mtrie[] old_table = next;
+            assert (new_min > min || new_max < min + count - 1);
+            assert (new_min >= min);
+            assert (new_max <= min + count - 1);
+            assert (new_max - new_min + 1 < count);
+            count = new_max - new_min + 1;
+            next = new Mtrie[count];
+
+            System.arraycopy(old_table, (new_min - min), next, 0, count);
+
+            min = new_min;
+        }
+        return true;
+    }
+
+    //  Remove specific subscription from the trie. Return true is it was
+    //  actually removed rather than de-duplicated.
+    public boolean rm (byte[] prefix_, int start_, Pipe pipe_)
+    {
+        return rm_helper (prefix_, start_,  pipe_);
+    }
+
+
+    private boolean rm_helper (byte[] prefix_, int start_, Pipe pipe_)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 21. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 257-266 into a separate method.